nsv13/code/modules/munitions/energy_weapons/laser_cannon.dm

```diff
@@ -0,0 +1,209 @@
+#define STATE_DISCONNECTED 1
+#define STATE_OFF 2
+#define STATE_CHARGING 3
+#define STATE_READY 4
+#define STATE_FIRING 5
+
+/*
+ * A ship-to-ship laser cannon that charges by drawing a large amount of power from the ship's grid.
+ * Can be researched by science and built by the crew.
+ */
+/obj/structure/ship_weapon/laser_cannon
+	name = "NT-WMG2 Laser cannon"
+	desc = "An experimental ship-to-ship energy weapon which fires a powerful destructive blast."
+	icon = 'nsv13/icons/obj/laser_cannon.dmi'
+	icon_state = "laser_cannon"
+
+	pixel_y = 0
+	pixel_x = 0
+	bound_width = 64
+	bound_height = 32
+	dir = 4
+
+	fire_sound = 'sound/weapons/lasercannonfire.ogg'
+	var/obj/machinery/computer/ship/laser_cannon_computer/computer
+	var/obj/item/stock_parts/cell/laser_cannon/cell
+	var/obj/structure/cable/attached
+	state = STATE_OFF
+
+	var/projectile_type = /obj/item/projectile/beam/laser/heavylaser
+
+	// Variables used for construction and deconstruction, copied from _machinery.dm
+	var/list/component_parts = null
+	var/panel_open = FALSE
+	var/obj/item/circuitboard/machine/laser_cannon/circuit // Circuit to be created and inserted when the machinery is created
+
+/*
+ * Overmap projectile
+ */
+/obj/item/projectile/bullet/laser
+	icon = 'icons/obj/projectiles.dmi'
+	icon_state = "heavylaser"
+	damage = 250
+
+/*
+ * Make sure we have a wire to drain power from, a circuitboard, and all the required parts for construction/deconstruction.
+ */
+/obj/structure/ship_weapon/laser_cannon/Initialize()
+	..()
+
+	attached = locate() in loc.contents // I don't know why I had to specify this but regular locate didn't work
+
+	if(!attached)
+		state = STATE_DISCONNECTED // Otherwise STATE_OFF as in the initial declaration
+
+	//This is duplicate _machinery.dm code
+	circuit = new /obj/item/circuitboard/machine/laser_cannon
+	circuit.apply_default_parts(src)
+
+	// Set charge to 0 for brand new cannons (i.e. map-placed cannons)
+	cell = locate(/obj/item/stock_parts/cell/laser_cannon) in component_parts
+	cell.charge = 0
+
+/*
+ * Destroys the cannon and its components. Copied from _machinery.dm
+ */
+/obj/structure/ship_weapon/laser_cannon/Destroy()
+	//This is duplicate _machinery.dm code
+	if(length(component_parts))
+		for(var/atom/A in component_parts)
+			qdel(A)
+		component_parts.Cut()
+	return ..()
```

Comment:
if this is duplicate code and defined on the superclass, and youre returning what the superclass returns, why are you copying the code from the superclass?

Reply:
_machinery.dm isn't the superclass because you made everything a structure

Reply:
GOD WHY THAT IS SO AWFUL